libnetwork/testutils/l3_segment_linux.go

```diff
@@ -0,0 +1,342 @@
+package testutils
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"net/netip"
+	"os"
+	"runtime"
+	"syscall"
+	"testing"
+
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"gotest.tools/v3/assert"
+)
+
+// Disposable represents any disposable resource that has a Destroy() method.
+type Disposable interface {
+	Destroy(Logger)
+}
+
+type tearDownFn func(Logger)
+
+func (fn tearDownFn) Destroy(logger Logger) {
+	fn(logger)
+}
+
+// DisposableEnv represents a common set of Disposable resources. When a DisposableEnv is destroyed, all its
+// resources are destroyed too.
+type DisposableEnv []Disposable
+
+func (env *DisposableEnv) Append(res Disposable) {
+	if env == nil {
+		*env = make([]Disposable, 1)
+	}
+	*env = append(*env, res)
+}
+
+func (env *DisposableEnv) Destroy(logger Logger) {
+	for i := len(*env) - 1; i >= 0; i-- {
+		(*env)[i].Destroy(logger)
+	}
+}
+
+// Netns represents a network namespace that might be named, and thus mounted under /run/netns.
+type Netns struct {
+	handle netns.NsHandle
+	name   string
+}
+
+// CurrentNetns returns an unnamed Netns for the current network namespace this thread is living in. If it fails to
+// get a handle for the current net namespace, it bails out.
+func CurrentNetns(t *testing.T) Netns {
+	t.Helper()
+
+	curNs, err := netns.Get()
+	assert.NilError(t, err)
+
+	return Netns{handle: curNs}
+}
+
+// NewNamedNetns creates a new named Netns and then switch back to the original net namespace. It bails out if it fails
+// in the process.
+func NewNamedNetns(t *testing.T, prefix string) Netns {
+	t.Helper()
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	origNs, err := netns.Get()
+	assert.NilError(t, err)
+
+	newNs := Netns{name: fmt.Sprintf("%s-%8x", prefix, rand.Uint32())}
+	newNs.handle, err = netns.NewNamed(newNs.name)
+	assert.NilError(t, err)
+
+	err = netns.Set(origNs)
+	assert.NilError(t, err)
+
+	return newNs
+}
+
+// Set joins the current goroutine to this Netns. The caller has to make sure the OS thread is locked.
+func (ns Netns) Set() error {
+	if !ns.handle.IsOpen() {
+		return nil
+	}
+	if err := netns.Set(ns.handle); err != nil {
+		return fmt.Errorf("could not switch to netns %s: %w", err)
+	}
+	return nil
+}
+
+// InNetns executes the provided fn in the Netns and then switch back to the original netns.
+func (ns Netns) InNetns(t *testing.T, fn func() error) error {
+	t.Helper()
+
+	orig := CurrentNetns(t)
+	runtime.LockOSThread()
+	defer func() {
+		runtime.UnlockOSThread()
+		if err := netns.Set(orig.handle); err != nil {
+			t.Logf("could not switch back to the original netns: %w", err)
+		}
```

Comment:
1. The order of operations is backwards. Restore, _then_ unlock.
2. Failing to restore the thread's network namespace is fatal for the thread, and makes any further use of the goroutine untrustworthy.
3. The `%w` format specifier only works with `fmt.Errorf`, not string formatters.

Since each test runs in a separate goroutine, we can prevent the failure to restore the thread's network namespace from affecting other tests by keeping the goroutine locked to the thread.

```suggestion
		if err := netns.Set(orig.handle); err != nil {
			t.Errorf("could not switch back to the original netns: %v", err)  // I'd rather use t.Fatalf but that can only be called from the test goroutine
			return
		}
		runtime.UnlockOSThread()
```

Reply:
> Failing to restore the thread's network namespace is fatal for the thread, and makes any further use of the goroutine untrustworthy.

Mh right. I think it'd be okay to call `t.Fatalf` here as you suggested, since `InNetns()` is called from the test goroutine. But that might not be future-proof :thinking: 

Other mistakes come from many refactorings (ie. disjoint defers, error being wrapped and returned instead of being logged); they'll be fixed next time I update this PR.